File: app/src/main/java/com/rahul/jhakadi/Authentications/SignInActivity.java
package com.rahul.jhakadi.Authentications;

import androidx.annotation.NonNull;
import androidx.appcompat.app.AppCompatActivity;

import android.content.Intent;
import android.os.Bundle;
import android.view.View;

import com.google.android.gms.tasks.OnCompleteListener;
import com.google.android.gms.tasks.Task;
import com.google.firebase.auth.AuthResult;
import com.google.firebase.auth.FirebaseAuth;
import com.rahul.jhakadi.MainActivity;
import com.rahul.jhakadi.R;
import com.rahul.jhakadi.Util;
import com.rahul.jhakadi.databinding.ActivitySignInBinding;

public class SignInActivity extends AppCompatActivity {

    private FirebaseAuth mAuth;
    private ActivitySignInBinding binding;
    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        binding = ActivitySignInBinding.inflate(getLayoutInflater());
        setContentView(binding.getRoot());
        mAuth = FirebaseAuth.getInstance();

        binding.signupbutton.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                Intent mainactivity= new Intent(getApplicationContext(), CreateAccountActivity.class);
                startActivity(mainactivity);
                finish();
            }
        });

        binding.signinbutton.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                final String mail= binding.userNameEdit.getText().toString();
                final String Password= binding.passwordNameEdit.getText().toString();

                if(mail.isEmpty() || Password.isEmpty()){
                    new Util().showMessage(SignInActivity.this,"Please Verify All Field");
                }
                else
                {
                    signIn(mail,Password);

                }
            }
        });

    }

    private void signIn(String mail, String Password) {
        mAuth.signInWithEmailAndPassword(mail,Password).addOnCompleteListener(new OnCompleteListener<AuthResult>() {
            @Override
            public void onComplete(@NonNull Task<AuthResult> task) {

                if (task.isSuccessful()){
                    Intent intent=new Intent(SignInActivity.this, MainActivity.class);
                    startActivity(intent);
                    finish();
                }
                else
                {
                    new Util().showMessage(SignInActivity.this,task.getException().getMessage());
                }

            }
        });
    }
}

File: app/src/main/java/com/rahul/jhakadi/Mvvm/Ui/SignInActivity.kt
package com.rahul.jhakadi.Mvvm.Ui

import android.content.Intent
import android.os.Bundle
import androidx.appcompat.app.AppCompatActivity
import androidx.lifecycle.Observer
import androidx.lifecycle.ViewModelProvider
import com.rahul.jhakadi.MainActivity
import com.rahul.jhakadi.Mvvm.Authentication.AuthRepository
import com.rahul.jhakadi.Mvvm.Authentication.AuthViewModel
import com.rahul.jhakadi.Mvvm.Authentication.AuthViewModelFactory
import com.rahul.jhakadi.Util
import com.rahul.jhakadi.databinding.ActivitySignInBinding

class SignInActivity : AppCompatActivity() {
    private lateinit var viewModel: AuthViewModel
    private var binding: ActivitySignInBinding? = null
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding = ActivitySignInBinding.inflate(layoutInflater)
        setContentView(binding!!.root)
        val repository = AuthRepository()
        val factory = AuthViewModelFactory(repository)
        viewModel = ViewModelProvider(this, factory).get(AuthViewModel::class.java)
        clickListener()
    }

    fun clickListener(){

        viewModel.signInStatus.observe(this, Observer { isSuccessful ->
            if (isSuccessful) {
                Util().showMessage(this,"Sign-IN successful!")
                // Navigate to the next screen or perform other actions as needed
                val intent = Intent(this@SignInActivity, MainActivity::class.java)
                startActivity(intent)
                finish()
            } else {
                // Handle sign-up failure if needed
                Util().showMessage(this,"Sign-up failure!")
            }
        })

        binding!!.signinbutton.setOnClickListener {
            viewModel.signIn(binding!!.userNameEdit.text.toString(), binding!!.passwordNameEdit.text.toString())
        }

        // Set up click listener for forgot password
        binding!!.forgotpassword.setOnClickListener {
            viewModel.resetPassword(binding!!.userNameEdit.text.toString())
        }

        // Observe the reset password status
        viewModel.resetPasswordStatus.observe(this, Observer { isSuccessful ->
            if (isSuccessful) {
                Util().showMessage(this,"Password reset email sent. Check your inbox.")
            } else {
                Util().showMessage(this,"Failed to send password reset email.")
            }
        })

        // Set up click listener for SignUp fragment
        binding!!.signupbutton.setOnClickListener {
            val intent = Intent(this, CreateAccountActivity::class.java)
            startActivity(intent)
            finish()
        }
    }

}